File: src/main/java/cs/android/CSApplication.java
package cs.android;

import android.app.Application;

import static cs.java.lang.CSLang.info;
import static cs.java.lang.CSLang.model;

public class CSApplication extends Application {

    private static CSApplication application;

    public CSApplication() {
        application = this;
    }

    public static CSApplication application() {
        return application;
    }

    public void onLowMemory() {
        super.onLowMemory();
        info("onLowMemory");
        model().logger().onLowMemory();
    }

    public void onTerminate() {
        super.onTerminate();
        info("onTerminate");
    }
}

File: library/renetik-android/src/main/java/renetik/android/model/CSApplication.kt
package renetik.android.model

import android.app.Application
import android.os.Environment.getExternalStorageDirectory
import renetik.android.lang.CSLog.logInfo
import renetik.android.lang.CSLogger
import renetik.android.lang.initializeHandler
import java.io.File

lateinit var application: CSApplication

abstract class CSApplication : Application() {

    abstract val name: String
    abstract val logger: CSLogger
    open val store: CSValueStore by lazy { CSValueStore("ApplicationSettings") }
    abstract val isDebugBuild: Boolean
    val externalFilesDir: File get() = getExternalFilesDir(null) ?: getExternalStorageDirectory()

    override fun onCreate() {
        super.onCreate()
        initializeHandler()
        application = this
    }

    override fun onLowMemory() {
        super.onLowMemory()
        logInfo("onLowMemory")
        logger.onLowMemory()
    }

    override fun onTerminate() {
        super.onTerminate()
        logInfo("onTerminate")
    }

}
